src/Mvc/Mvc.Core/src/ModelBinding/Binders/ComplexObjectModelBinder.cs

```diff
@@ -0,0 +1,754 @@
+            string fieldName,
+            string modelName)
+        {
+            Debug.Assert(parameter.MetadataKind == ModelMetadataKind.Parameter);
+
+            ModelBindingResult result;
+            using (bindingContext.EnterNestedScope(
+                modelMetadata: parameter,
+                fieldName: fieldName,
+                modelName: modelName,
+                model: null))
+            {
+                await parameterBinder.BindModelAsync(bindingContext);
+                result = bindingContext.Result;
+            }
+
+            if (!result.IsModelSet && parameter.IsBindingRequired)
+            {
+                var message = parameter.ModelBindingMessageProvider.MissingBindRequiredValueAccessor(fieldName);
+                bindingContext.ModelState.TryAddModelError(modelName, message);
+            }
+
+            return result;
+        }
+
+        internal int CanCreateModel(ModelBindingContext bindingContext)
+        {
+            var isTopLevelObject = bindingContext.IsTopLevelObject;
+
+            // If we get here the model is a complex object which was not directly bound by any previous model binder,
+            // so we want to decide if we want to continue binding. This is important to get right to avoid infinite
+            // recursion.
+            //
+            // First, we want to make sure this object is allowed to come from a value provider source as this binder
+            // will only include value provider data. For instance if the model is marked with [FromBody], then we
+            // can just skip it. A greedy source cannot be a value provider.
+            //
+            // If the model isn't marked with ANY binding source, then we assume it's OK also.
+            //
+            // We skip this check if it is a top level object because we want to always evaluate
+            // the creation of top level object (this is also required for ModelBinderAttribute to work.)
+            var bindingSource = bindingContext.BindingSource;
+            if (!isTopLevelObject && bindingSource != null && bindingSource.IsGreedy)
+            {
+                return NoDataAvailable;
+            }
+
+            // Create the object if:
+            // 1. It is a top level model.
+            if (isTopLevelObject)
+            {
+                return ValueProviderDataAvailable;
+            }
+
+            // 2. Any of the model properties can be bound.
+            return CanBindAnyModelItem(bindingContext);
+        }
+
+        private int CanBindAnyModelItem(ModelBindingContext bindingContext)
+        {
+            // If there are no properties on the model, and no constructor parameters, there is nothing to bind. We are here means this is not a top
+            // level object. So we return false.
+            var modelMetadata = bindingContext.ModelMetadata;
+            var performsConstructorBinding = bindingContext.Model == null && modelMetadata.BoundConstructor != null;
```

Comment:
Again this violates the principle of least surprise, especially when the `record` is a property of something else. We replace `struct`s even if they have an existing value, why not `record`s❔ What happens when devs use `with` to attempt to create a new `record` with new values that are passed to the constructor e.g. does it fail or does it use the copy constructor or clone method then overwrite the values❔

Reply:
You get a new instance though. TryUpdateModel scenarios require that you update the model in place. For record types this would require us to set parameter values via properties, which feels iffy to me. But let's discuss this with the larger team, maybe they feel differently.

Reply:
Yup, getting a new instance is exactly what I'd expect if I have
``` c#
public record Person(string Name);
```
or
```
public record Person
{
    public string Name { get; init; }
}
```
as long as I'm calling `TryUpdateModel(...)` on a class containing `Person`.

Reply:
I added some tests and it admittedly feels wonky. Maybe a fix for this is to always bind the constructor unless it's a top-level type.

Reply:
Filed https://github.com/dotnet/aspnetcore/issues/24265 for tracking